src/modules/flow-metatype/http-composed-server/http-composed-server.c

```diff
@@ -0,0 +1,1126 @@
+                .description = "The URL used on requests",
+                .data_type = "string",
+                .required = false,
+                .offset = offsetof(struct http_composed_server_options, path),
+                .size = sizeof(const char *),
+            },
+            {
+                .name = "port",
+                .description = "The port where the server will bind",
+                .data_type = "int",
+                .required = false,
+                .offset = offsetof(struct http_composed_server_options, port),
+                .size = sizeof(int),
+            },
+            {},
+        },
+    }),
+};
+
+static int
+setup_description(struct http_composed_server_type *type)
+{
+    struct sol_flow_node_type_description *desc;
+    struct sol_flow_port_description **p;
+    struct http_composed_server_port_in *port_type_in;
+    struct http_composed_server_port_out *port_type_out;
+    int i, j;
+
+    type->base.description = calloc(1, sizeof(struct sol_flow_node_type_description));
+    SOL_NULL_CHECK(type->base.description, -ENOMEM);
+
+    type->base.description = memcpy((struct sol_flow_node_type_description *)type->base.description,
+        &sol_flow_node_type_http_server_composed_description, sizeof(struct sol_flow_node_type_description));
+
+    desc = (struct sol_flow_node_type_description *)type->base.description;
+
+    /* Extra slot for NULL sentinel at the end. */
+    desc->ports_in = calloc(type->ports_in.len + 1, sizeof(struct sol_flow_port_description *));
+    SOL_NULL_CHECK_GOTO(desc->ports_in, fail_ports_in);
+
+    p = (struct sol_flow_port_description **)desc->ports_in;
+    for (i = 0; i < type->ports_in.len; i++) {
+        p[i] = calloc(1, sizeof(struct sol_flow_port_description));
+        SOL_NULL_CHECK_GOTO(p[i], fail_ports_in_desc);
+
+        port_type_in = sol_vector_get(&type->ports_in, i);
+
+        p[i]->name = port_type_in->name;
+        p[i]->description = "Input port";
+        p[i]->data_type = port_type_in->base.packet_type->name;
+        p[i]->array_size = 0;
+        p[i]->base_port_idx = i;
+        p[i]->required = false;
+    }
+
+    /* Extra slot for NULL sentinel at the end. */
+    desc->ports_out = calloc(type->ports_out.len + 1, sizeof(struct sol_flow_port_description *));
+    SOL_NULL_CHECK_GOTO(desc->ports_out, fail_ports_in_desc);
+
+    p = (struct sol_flow_port_description **)desc->ports_out;
+    for (j = 0; j < type->ports_out.len; j++) {
+        p[j] = calloc(1, sizeof(struct sol_flow_port_description));
+        SOL_NULL_CHECK_GOTO(p[j], fail_ports_out_desc);
+
+        port_type_out = sol_vector_get(&type->ports_out, j);
+
+        p[j]->name = port_type_out->name;
+        p[j]->description = "Output port";
+        p[j]->data_type = port_type_out->base.packet_type->name;
+        p[j]->array_size = 0;
+        p[j]->base_port_idx = j;
+        p[j]->required = false;
+    }
+
+    return 0;
+
+fail_ports_out_desc:
+    if (j > 0) {
```

Comment:
This looks strange. Say you failed on the second iteration above, you start freeing index one here.